cdap-data-fabric/src/main/java/co/cask/cdap/store/OwnerTable.java

```diff
@@ -0,0 +1,229 @@
+   * @throws IOException is thrown when there is an error writing to the table.
+   * @throws AlreadyExistsException is thrown when the key entityId already exists.
+   */
+  public void add(final NamespacedEntityId entityId,
+                  final KerberosPrincipalId kerberosPrincipalId) throws IOException, AlreadyExistsException {
+    Optional<StructuredRow> row = table.read(ImmutableList.of
+        (Fields.stringField(StoreDefinition.OwnerStore.PRINCIPAL_FIELD, createRowKey(entityId))));
+    if (row.isPresent()) {
+      throw new AlreadyExistsException(entityId,
+          String.format("Owner information already exists for entity '%s'.",
+              entityId));
+    }
+    Field<String> principalField = Fields.stringField(StoreDefinition.OwnerStore.PRINCIPAL_FIELD,
+                                                      createRowKey(entityId));
+    Field<byte[]> keytabField = Fields.bytesField(StoreDefinition.OwnerStore.KEYTAB_FIELD,
+        Bytes.toBytes(kerberosPrincipalId.getPrincipal()));
+    table.upsert(ImmutableList.of(principalField, keytabField));
+  }
+
+  /**
+   * Function that checks if a given entity id exists.
+   *
+   * @param entityId NamespacedEntityId to check for existence.
+   * @return true if the entityId exists, false otherwise.
+   * @throws IOException is thrown when there is an error reading from the table.
+   */
+  public boolean exists(final NamespacedEntityId entityId) throws IOException {
+    Optional<StructuredRow> row = table.read(ImmutableList.of
+        (Fields.stringField(StoreDefinition.OwnerStore.PRINCIPAL_FIELD, createRowKey(entityId))));
+    return row.isPresent();
+  }
+
+  /**
+   * Get Kerberos principal for a given entity id
+   *
+   * @param entityId NamespacedEntityId for the requested KerberosPrincipalId.
+   * @return KerberosPrincipalId for the given entityId.
+   * @throws IOException is thrown when there is an error reading from the table.
+   */
+  @Nullable
+  public KerberosPrincipalId getOwner(final NamespacedEntityId entityId) throws IOException {
+    Optional<StructuredRow> row = table.read(ImmutableList.of
+        (Fields.stringField(StoreDefinition.OwnerStore.PRINCIPAL_FIELD, createRowKey(entityId))));
+
+    return row.isPresent() ?
+        new KerberosPrincipalId(Bytes.toString(row.get().getBytes(StoreDefinition.OwnerStore.KEYTAB_FIELD))) : null;
+  }
+
+  /**
+   * Delete the entityId and corresponding kerberos principalId.
+   *
+   * @param entityId NamespacedEntityId to delete.
+   * @throws IOException is thrown when there is an error deleting the entry.
+   */
+  public void delete(final NamespacedEntityId entityId) throws IOException {
+    table.delete(ImmutableList.of(Fields.stringField(StoreDefinition.OwnerStore.PRINCIPAL_FIELD,
+                                                     createRowKey(entityId))));
+  }
+
+  //Create row key from Namespaced Entity ID
+  private String createRowKey(NamespacedEntityId namespacedEntityId) {
```

Comment:
Here I meant to have recursive calls back to parent which will avoid the repetition like this: 

```
  private String createRowKey(NamespacedEntityId namespacedEntityId) {
    StringBuilder builder = new StringBuilder();
    createRowKeyHelper(namespacedEntityId, builder);
    return builder.toString();
  }

  private void createRowKeyHelper(NamespacedEntityId namespacedEntityId, StringBuilder builder) {
    EntityType type = namespacedEntityId.getEntityType();
    switch (type) {
      case NAMESPACE:
        NamespaceId namespaceId = (NamespaceId) namespacedEntityId;
        builder.append(NAMESPACE_ROW_KEY_PREFIX);
        builder.append(ROW_KEY_SEPARATOR);
        builder.append(namespaceId.getNamespace());
        return;
      case DATASET:
        DatasetId datasetId = (DatasetId) namespacedEntityId;
        createRowKeyHelper(datasetId.getParent(), builder);

        builder.append(ROW_KEY_SEPARATOR);
        builder.append(DATASET_ID_ROW_KEY_PREFIX);
        builder.append(ROW_KEY_SEPARATOR);
        builder.append(datasetId.getDataset());
        return;
      case ARTIFACT:
        ArtifactId artifactId = (ArtifactId) namespacedEntityId;
        createRowKeyHelper(artifactId.getParent(), builder);
        builder.append(ROW_KEY_SEPARATOR);

        builder.append(ROW_KEY_SEPARATOR);
        builder.append(ARTIFACT_ID_ROW_KEY_PREFIX);
        builder.append(ROW_KEY_SEPARATOR);
        builder.append(artifactId.getArtifact());

        builder.append(ROW_KEY_SEPARATOR);
        builder.append(ARTIFACT_VERSION_ROW_KEY_PREFIX);
        builder.append(ROW_KEY_SEPARATOR);
        builder.append(artifactId.getVersion());
        return;
      case APPLICATION:
        ApplicationId application = (ApplicationId) namespacedEntityId;
        createRowKeyHelper(application.getParent(), builder);
        builder.append(ROW_KEY_SEPARATOR);

        builder.append(APP_ID_ROW_KEY_PREFIX);
        builder.append(ROW_KEY_SEPARATOR);
        builder.append(application.getApplication());

        builder.append(ROW_KEY_SEPARATOR);
        builder.append(ARTIFACT_VERSION_ROW_KEY_PREFIX);
        builder.append(ROW_KEY_SEPARATOR);
        builder.append(application.getVersion());
        return;
      case PROGRAM:
        ProgramId programId = (ProgramId) namespacedEntityId;
        createRowKeyHelper(programId.getParent(), builder);
        builder.append(ROW_KEY_SEPARATOR);
        builder.append(PROGRAM_TYPE_ROW_KEY_PREFIX);
        builder.append(ROW_KEY_SEPARATOR);
        builder.append(programId.getType());

        builder.append(ROW_KEY_SEPARATOR);
        builder.append(PROGRAM_ID_ROW_KEY_PREFIX);
        builder.append(ROW_KEY_SEPARATOR);
        builder.append(programId.getProgram());
        return;
      case SCHEDULE:
        ScheduleId scheduleId = (ScheduleId) namespacedEntityId;
        createRowKeyHelper(scheduleId.getParent(), builder);
        
        builder.append(ROW_KEY_SEPARATOR);
        builder.append(SCHEDULE_NAME_ROW_KEY_PREFIX);
        builder.append(ROW_KEY_SEPARATOR);
        builder.append(scheduleId.getSchedule());
        return;
      default:
        throw new IllegalArgumentException("Unexpected.");
    }
  }
```

Reply:
Not using a recursive approach for maintainability.